### PR TITLE
Add S3 Select pushdown for JSON files

### PR DESCRIPTION
**Pull request by dnanuti**
Add JSON support for Hive connector with S3 Select pushdown

Similarly to CSV, enable S3 Select pushdown for Hive tables relying on JSON files stored in S3 and serialized/deserialized with `org.apache.hive.hcatalog.data.JsonSerDe`.
Small refactoring for IonSqlQueryBuilder needed to support query generation for JSON.
The pushdown logic works for base columns, aligned with CSV implementation.

Test plan - Added Docker tests 

Locally tested:
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.facebook.presto.hive.s3select.TestHiveFileSystemS3SelectJsonPushdown
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.apache.hadoop.fs.HadoopExtendedFileSystemCache (file:/Users/dnnanuti/.m2/repository/com/facebook/presto/presto-hive-common/0.280-SNAPSHOT/presto-hive-common-0.280-SNAPSHOT.jar) to field java.lang.reflect.Field.modifiers
WARNING: Please consider reporting this to the maintainers of org.apache.hadoop.fs.HadoopExtendedFileSystemCache
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
2023-01-22T03:01:48.848-0600 WARNING Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2023-01-22T03:01:48.932-0600 INFO Successfully loaded & initialized native-bzip2 library system-native
2023-01-22T03:01:48.936-0600 INFO Successfully loaded & initialized native-zlib library
2023-01-22T03:01:50.846-0600 WARNING NoSuchMethodException was thrown when disabling normalizeUri. This indicates you are using an old version (< 4.5.8) of Apache http client. It is recommended to use http client version >= 4.5.9 to avoid the breaking change introduced in apache client 4.5.7 and the latency in exception handling. See https://github.com/aws/aws-sdk-java/issues/1919 for more information
2023-01-22T03:01:52.067-0600 INFO io.bytes.per.checksum is deprecated. Instead, use dfs.bytes-per-checksum
2023-01-22T03:01:52.445-0600 INFO Got brand-new decompressor [.bz2]
2023-01-22T03:01:52.670-0600 INFO Got brand-new decompressor [.gz]
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.927 s - in com.facebook.presto.hive.s3select.TestHiveFileSystemS3SelectJsonPushdown
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  27.556 s
[INFO] Finished at: 2023-01-22T09:01:53Z
[INFO] ------------------------------------------------------------------------
```

```
== RELEASE NOTES ==

Hive Changes
* Add Amazon S3 Select pushdown for JSON files.
```

### CODE REVIEW

**Comment by dnanuti** [comment]
@pettyjamesm

File: presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/S3SelectTestHelper.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    private CacheConfig cacheConfig;
+    private MetastoreClientConfig metastoreClientConfig;
+
+    @BeforeClass
+    public void setUp()
+    {
+        executor = newCachedThreadPool(daemonThreadsNamed("hive-%s"));
+    }
+
+    @AfterClass(alwaysRun = true)
```
Comment (pettyjamesm) [inline presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/S3SelectTestHelper.java:106]: These annotations only work when run on the current class as a test. This appears to just be a utility class used from other tests, so those will need to have setup / teardown methods of their own with these annotations in order to be called.

File: presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/TestHiveFileSystemS3SelectJsonPushdown.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            "hive.hadoop2.metastoreHost",
+            "hive.hadoop2.metastorePort",
+            "hive.hadoop2.databaseName",
+            "hive.hadoop2.s3.awsAccessKey",
+            "hive.hadoop2.s3.awsSecretKey",
+            "hive.hadoop2.s3.writableBucket"})
+    @BeforeClass
+    public void setup(String host, int port, String databaseName, String awsAccessKey, String awsSecretKey, String writableBucket)
+    {
+        s3SelectTestHelper = new S3SelectTestHelper(host, port, databaseName, awsAccessKey, awsSecretKey, writableBucket, true);
```
Comment (pettyjamesm) [inline presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/TestHiveFileSystemS3SelectJsonPushdown.java:51]: Missing `s3SelectTestHelper.setUp()` call. Does this test pass in the current state?

**Review by pettyjamesm — COMMENTED** [review]
Added some comments about issues I see with the test utility and usage, otherwise the main changes look good to me

File: presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/S3SelectTestHelper.java (hunk 1 above)
Comment (dnanuti) [reply presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/S3SelectTestHelper.java:106]: Correct, thanks for catching this! I mixed things up during rebasing & moving from the test class to the helper class.

File: presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/TestHiveFileSystemS3SelectJsonPushdown.java (hunk 2 above)
Comment (dnanuti) [reply presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/TestHiveFileSystemS3SelectJsonPushdown.java:51]: Yes, the test passes:
image

Comment (dnanuti) [reply presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/TestHiveFileSystemS3SelectJsonPushdown.java:51]: The executor is also instantiated in the constructor and that is actually passed on as a parameter, so it did not depend on the setup method, it just does not release the resources explicitly.

Comment (dnanuti) [reply presto-hive-hadoop2/src/test/java/com/facebook/presto/hive/s3select/TestHiveFileSystemS3SelectJsonPushdown.java:52]: I ditched the setup method in utility class and updated the tests teardown. Nice catch! 🥇

**Review by pettyjamesm — APPROVED** [review]
Looks good to me, let's see if we can get a green build- the jenkins job seemed to get terminated unexpectedly for some unknown (but assumed to be unrelated) reason.

**Review by pettyjamesm — APPROVED** [review]
Approved, thanks for the contribution!

**Pull request merged**